syntax = "proto3";

package ecliptix.proto.protocol;
option csharp_namespace = "Ecliptix.Protobuf.Protocol";

enum PubKeyExchangeState {
  INIT = 0;
  COMPLETE = 1;
  PENDING = 2;
  FAILED = 3;
}

enum PubKeyExchangeType {
  INITIAL_HANDSHAKE = 0;
  DATA_CENTER_EPHEMERAL_CONNECT = 1;
  SERVER_STREAMING = 2;
  DEVICE_TO_DEVICE = 3;
}

message PubKeyExchange {
  PubKeyExchangeState state = 1;
  PubKeyExchangeType of_type = 2;
  bytes payload = 3;
  bytes initial_Dh_public_Key = 4;
}

message PublicKeyBundle {
  bytes identity_public_key = 1;

  uint32 signed_pre_key_id = 2;
  bytes signed_pre_key_public_key = 3;
  bytes signed_pre_key_signature = 4;

  message OneTimePreKey {
    uint32 pre_key_id = 1;
    bytes public_key = 2;
  }
  repeated OneTimePreKey one_time_pre_keys = 5;

  bytes identity_x25519_public_key = 6;

  bytes ephemeral_x25519_public_key = 7;

  // Post-quantum Kyber-768 public key (1184 bytes). Always required in hybrid mode.
  bytes kyber_public_key = 8;

  // Optional Kyber ciphertext for initial encapsulation (1088 bytes).
  // Present when the initiator encapsulates to the responder's Kyber key as part of the handshake.
  bytes kyber_ciphertext = 9;
}
